sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -236,6 +284,211 @@ private Session session() {
         return result;
     }
 
+    private CompletableFuture<Long> executeAlterTableChangeNumberOfShards(AlterTableAnalyzedStatement analysis) {
+        final DocTableInfo table = analysis.table();
+        final boolean isPartitioned = table.isPartitioned();
+        String sourceIndexName, sourceIndexAlias;
+        if (isPartitioned) {
+            Optional<PartitionName> partitionName = analysis.partitionName();
+            assert partitionName.isPresent() : "Resizing operations for partitioned tables " +
+                                               "are only supported at partition level";
+            sourceIndexName = partitionName.get().asIndexName();
+            sourceIndexAlias = table.ident().indexName();
+        } else {
+            sourceIndexName = table.ident().indexName();
+            sourceIndexAlias = null;
+        }
+        String targetIndexName = SHRINK_PREFIX + sourceIndexName;
+
+        int targetNumberOfShards = analysis.tableParameter().settings().getAsInt(SETTING_NUMBER_OF_SHARDS, 0);
+        validateForResize(sourceIndexName, targetNumberOfShards);
+
+        List<ChainableAction<Long>> actions = new ArrayList<>(7);
+
+        // pickup node
+        DiscoveryNode resizeNode = getNodeForResize();
+
+        // change allocation
+        actions.add(new ChainableAction<>(
+            () -> restrictAllocationToNode(resizeNode.getName(), sourceIndexName),
+            () -> revertAllocation(sourceIndexName)
+        ));
+
+        // shrink index
+        actions.add(new ChainableAction<>(
+            () -> shrinkIndex(sourceIndexName, sourceIndexAlias, targetIndexName, targetNumberOfShards),
+            () -> deleteIndex(targetIndexName)
+        ));
+
+        // close index
+        actions.add(new ChainableAction<>(
+            () -> closeTable(sourceIndexName, targetIndexName),
+            () -> openTable(sourceIndexName, targetIndexName)
+        ));
+
+        // index rename
+        final String backupIndexName = ".backup" + sourceIndexName;
+        String[] renameSourceNames = new String[2];
+        String[] renameTargetNames = new String[2];
+        renameSourceNames[0] = sourceIndexName;
+        renameTargetNames[0] = backupIndexName;
+        renameSourceNames[1] = targetIndexName;
+        renameTargetNames[1] = sourceIndexName;
+
+        String[] renameSourceNamesReverse = new String[2];
+        String[] renameTargetNamesReverse = new String[2];
+        renameTargetNamesReverse[0] = renameSourceNames[1];
```

Comment:
maybe a personal preference, but this is very hard to follow, I'd rather have the names of the indexes be assigned (like above for `renameSourceNames` and `renameTargetNames`) and maybe extract this in a utility method that's unit tested

Reply:
I agree, it's one of the things that I do need to revise if we decide to keep this approach of renaming instead of exchange (in previous commit). What I don't like on this rename is that the operation is not symmetrical by nature so you need to do this processing in order to achieve the desired result. I was also considering using a different data structure to prepare the 'inversion' but I want to discuss first if it's worth going that way

Reply:
ok, we've decided to keep the BulcRename Operation thus the API will change - this was an init basic implementation until a decision was made